### ISSUE TITLE
[prefer-optional-chain] Ignore assignment patterns

### ISSUE THREAD

**Issue by NullDivision** [typescript-eslint/typescript-eslint#4892]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-optional-chain": "warn"
  }
}
```

```TS
// your repro code case
const a: { b: string | null } = { b: null };

const x = a.b && a.b?.includes('test');
```

**Expected Result**

The code is considered valid.

**Actual Result**

ESLint tells me `Prefer using an optional chain expression instead, as it's more concise and easier to read.`.

**Additional Info**

**Versions**

| package                            | version   |
| ---------------------------------- | --------- |
| `@typescript-eslint/eslint-plugin` | `5.21.0`  |
| `@typescript-eslint/parser`        | `5.21.0`  |
| `TypeScript`                       | `4.6.4`   |
| `ESLint`                           | `8.14.0`  |
| `node`                             | `16.15.0` |

In the above example the output of optional chaining would not be equivalent as a.b would short- circuit with a null value and return `null`. I'm sure this falls under the warning mentioned in the docs about edge cases but a nicer solution would be to either:

a. ignore optional chaining warnings for _assignments only_ (using an option for the rule)
b. use type information (mentioned in #4820) to determine if the resulting value might be null

**Comment by bradzacher** [comment]
> In the above example the output of optional chaining would not be equivalent as a.b would short- circuit with a null value and return null.

`a.b && a.b.thing` will return null if `a.b` is null
`a.b?.thing` will return undefined if `a.b` is null.

Either way they both short-circuit. They are mostly equivalent - as long as your code doesn't differentiate between null and undefined.

---

I don't think an option here has wide-enough applicability to add to the rule. Feel free to fork it if thus is important to you!

**Comment by JoshuaKGoldberg** [comment]
Note: per #4820 we're going to add type checking to the rule so that it can understand the difference between a `null` and `undefined` `a.b` in `a.b && a.b.thing` (Brad's comment just before this one).
